### ISSUE TITLE
[rush] Add support for locking versions of dependencies at publish time

### ISSUE THREAD

**Issue by dmichon-msft** [Microsoft/web-build-tools#1096]
When working in a high-traffic RUSH monorepo, adding a new project that contains references to other projects (or adding references to other within-repo projects to an existing project) becomes a chore of babysitting the pull request and manually fixing dependency versions to ensure that the correct version is targeted when the build job runs and the change to package.json gets committed.

We prefer to have our published package.json files refer to exact dependency versions for stability, but at development time, the use of linking makes this largely irrelevant.

I propose an addition to the publish flow that (if requested in config) locks the versions of within-repo dependencies (and possibly also of packages that are configured via common-versions) in the published package.json files, while leaving the checked-in versions with a wildcard specifier `*`. This would greatly simplify the process of adding within-repo dependencies to package.json files while still maintaining the version guarantee.

**Comment by ThomasMichon** [comment]
Looking at the logic for `PublishAction`, I see an opportunity to support this behavior if a flag is set in `rush.json`.

```json
{
  "storeLooseVersions": true
}
```

The above flag would tell Rush that during version bumping and publishing, packages should refer to each other by exact versions, but at all other times they should refer to each other using `*` when checked into source.

If the above flag is set, then during `PublishAction`, Rush can 'lock' and 'unlock' the values in `package.json` between publish phases, preserving all the existing change detection logic and flow mechanics:

1. Check out code, create temp branch
2. 'lock' all dependencies, modifying package.json files. *(New)*
3. Apply change files, incrementing versions. Since the versions are locked, updates cascade correctly.
4. 'unlock' all dependencies back to `*` between projects. *(New)*
5. Check for changed files (this way the 'lock' step is ignored for change detection)
6. Re-lock all package.json files *(New)*
7. Run publish flow on changed packages (the locking doesn't count)
8. 'unlock' all package.json files again *(New)*
9. Commit changes and tags
10. Push to target branch

There's a bit of churn involved, but by doing it this way we'd be able to avoid the commit churn caused by version bumps, while ensuring we consistently publish 'locked' packages as sets.

**Comment by iclanton** [comment]
This is an interesting solution to a problem I've definitely had to deal with in the past.

Some general feedback:
- I think this configuration option should go in specific version policies, not in `rush.json`
- Instead of `storeLooseVersions`, what do you think of the being the opposite? Something like `publishStrictDependencyVersions` or `publishExactDependencyVersions`?
- I don't think we should assume the committed versions will always be `*`. I think the algorithm should allow for any version selector.
- This should probably be considered a future addition, but I can see scenarios where we want to lock the published dependency versions to a particular non-exact specifier. I think for that scenario, we may want to just pursue the path of Rush generating and updating `package.json` files during `rush update` and avoid having them committed to git at all.

**Comment by dmichon-msft** [comment]
I agree with preserving the original specifiers from `package.json`, though we would then need to verify that the exact version satisfies the specifier, and error out (or potentially fix it) if it doesn't.

**Comment by ThomasMichon** [comment]
The problem with allowing 'original specifiers' is that Rush usually maintains these. So if you have some different range specifier than '*' in `package.json`, you don't need the feature proposed here since you want Rush to write back to `package.json`. `*` is the only range completely immune to the need to keep it updated.

**Comment by octogonz** [comment]
We have had a plan for a while to make the `package.json` file into an output file, rather than a source file tracked by Git. When `rush install` is invoked, it would read an input source file such as `package-rush.json` and then overwrite `package.json` with a generated copy. This would have the following benefits:

- Eliminate the merge conflicts that frequently occur when a CI job (e.g. `rush version` or XStitch) bumps versions in master, because the CI job would now only be modifying a file such as `version-policies.json` which humans never touch
- Provide a way to strip references to internal packages from `package.json` before publishing, and generally prettify the JSON
- Provide a way for "beta" and "production" publishing targets to adjust pointers such as `main` and `typings` in *package.json* without churning the Git history

@ThomasMichon I suspect your idea could be incorporated into this design as well.

**Comment by octogonz** [comment]
@nickpape-msft FYI

**Comment by ThomasMichon** [comment]
@octogonz Good to hear from you! I remember discussing this idea with you a bit back, but... circumstances... made me wonder if that was still in the cards.

When I was thinking about the `package-rush.json` 'stub' file, it occurred to me that if we moved most of the logic from `package.json` into yet another `.json` file, we'd lose all of the IDE support for that file. No auto-complete of available package names and versions, for example. So I started to think of ways to keep using `package.json` directly for this, and the above came out.

I noticed that some projects using Yarn or Lerna actually do the `*` thing, so there's even a precedent for this as being a 'valid' approach, even if it may not be the 'best' one.

**Comment by octogonz** [comment]
> When I was thinking about the `package-rush.json` 'stub' file, it occurred to me that if we moved most of the logic from `package.json` into yet another `.json` file, we'd lose all of the IDE support for that file. No auto-complete of available package names and versions, for example. So I started to think of ways to keep using `package.json` directly for this, and the above came out.

The IDE support would still work, since the `package.json` file would be a valid file that correctly describes the package.  However since the IntelliSense would be editing an output file, afterwards you'd need to copy the result into the input file `package-rush.json`.  That seemed like an okay experience to me.  If people think it would be awkward, please let us know.  

If IDE support is the **only** objection to this design, maybe we could call the file something more generic like `package-input.json` and then teach the VS Code language service to recognize that filename.

Another idea would be to provide a Rush command that detects manual changes to **package.json** (by diffing the actual contents with the generated content), and automatically update **package-rush.json**.

**Comment by ThomasMichon** [comment]
I don't strongly object to `package-input.json` or `package-rush.json`, though it does have the effect of finally coupling an actual single package project layout to Rush. Right now, one could still 'port' a project source from Rush to Yarn, or whatever, because all the conventions within a package are the same. Once we go down the synthetic-project route, things get interesting.

My proposal here was a way to side-step that eventual future for the moment, to douse the flames in a repo which experiences evergreen or PR stoppages caused by publishing a couple days a week!

**Comment by octogonz** [comment]
That rationale makes sense to me. I checked with @iclanton yesterday and he agreed.
